agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/telemetry/TelemetryChannel.java

```diff
@@ -234,115 +257,133 @@ private CompletableResultCode internalSend(
-                  statsbeatModule
-                      .getNetworkStatsbeat()
-                      .incrementRequestFailureCount(instrumentationKey);
-                }
-                // no need to write to disk again when failing to send raw bytes from the persisted
-                // file
-                if (!persisted) {
-                  writeToDiskOnFailure(byteBuffers, instrumentationKey);
-                }
-              }
-
-              if (!persisted) {
-                // persisted byte buffers don't come from the pool so shouldn't go back to the pool
-                byteBufferPool.offer(byteBuffers);
-              }
-              result.fail();
-            });
+            responseHandler(
+                instrumentationKey,
+                startTime,
+                () -> {
+                  onSuccess.run();
+                  result.succeed();
+                },
+                retryable -> {
+                  onFailure.accept(retryable);
+                  result.fail();
+                },
+                operationLogger),
+            errorHandler(
+                instrumentationKey,
+                retryable -> {
+                  onFailure.accept(retryable);
+                  result.fail();
+                },
+                operationLogger));
     return result;
   }
 
-  private void writeToDiskOnFailure(List<ByteBuffer> byteBuffers, String instrumentationKey) {
-    if (localFileWriter != null) {
-      localFileWriter.writeToDisk(byteBuffers, instrumentationKey);
+  private Consumer<HttpResponse> responseHandler(
+      String instrumentationKey,
+      long startTime,
+      Runnable onSuccess,
+      Consumer<Boolean> onFailure,
+      OperationLogger operationLogger) {
+
+    return response ->
+        response
+            .getBodyAsString()
+            .subscribe(
+                body -> {
+                  int statusCode = response.getStatusCode();
+                  switch (statusCode) {
+                    case 200: // SUCCESS
+                      operationLogger.recordSuccess();
+                      onSuccess.run();
+                      break;
+                    case 206: // PARTIAL CONTENT, Breeze-specific: PARTIAL SUCCESS
+                      operationLogger.recordFailure(
+                          getErrorMessageFromPartialSuccessResponse(body));
+                      onFailure.accept(false);
+                      break;
+                    case 408: // REQUEST TIMEOUT
+                    case 429: // TOO MANY REQUESTS
+                    case 500: // INTERNAL SERVER ERROR
+                    case 503: // SERVICE UNAVAILABLE
+                      operationLogger.recordFailure(
+                          "received response code "
+                              + statusCode
+                              + " (telemetry will be stored to disk and retried later)");
+                      onFailure.accept(true);
+                      break;
```

Comment:
this is the list of response codes that 2.x used to retry on, I think it's a good list, not sure when we diverged from it